### DISCUSSION TITLE
Logs de Erro retornando 0 em outlet 0

### DISCUSSION THREAD

**Discussion by emviveros** [charlesneimog/py4pd#91] [Q&A]
Olá, @charlesneimog !

Estou trabalhando em um projeto, e o comportamento padrão do py4pd de retornar 0 pelo outlet 0, nos logs de erro, está me causando problemas.

Posso criar mais dois outlets e trabalhar com os outlets 1 e 2 para as minhas saídas, mas isso seria contra intuitivo para um usuário comum de Pd, que por padrão tentaria encontrar as saídas a partir do outlet 0. Existe algum jeito de desabilitar esse comportamento?

**Comment by charlesneimog** [comment]
oi @emviveros, joia?

Estou supondo que você está usando a função `pd.add_object` para criar o seu objeto. Se sim, é so colocar `ignore_none_return=True`. Provalmente a função Python está retornando `None` e vc **não** está usando `py_out=True`, portanto o objeto converte os dados de Python para as representações equivalentes em Pd (quando possível). Nesse caso, `None` é representado por 0, assim como `False`. `True` é 1. 

Se isso não resolver, compartilhe um exemplo simples do patch, por que eu não consigo lembrar nenhum parte do código onde eu retorno 0 em caso de erro de execução.

Talvez eu não esteja entendendo o que está acontecendo.

**Reply by emviveros** [reply]
Obrigado por responder.

Fiz uma revisão aqui, e agora, o comportamento de mandar um 0 pelo outlet 0 quando recebo um None, ficou interessante para minha aplicação. 

De toda forma testei o `ignore_none=True` que está documentado em: https://py4pd.readthedocs.io/en/latest/python-users/pd.new_object/config/#ignore_none , e funcionou perfeitamente para minha dúvida também, pois mesmo minha função em python retornando None, a saída 0 no outlet 0 foi suprimida.

Porém fui fazer um último teste de adição de um objeto que deveria dar um output 0 quando recebe
 None e não funcionou, mesmo colocando ignore_none=False.

Segue o teste que fiz para facilitar o entendimento. 
[Outlet0None.zip](https://github.com/charlesneimog/py4pd/files/15139481/Outlet0None.zip)

**Reply by charlesneimog** [reply]
@emviveros  Hmmm, entendi o que está acontecendo. Isso ta confuso mesmo.


É que esse ignore none ele só funciona quanto `py_out=True`. Até coloquei um comentário sobre isso. 

https://github.com/charlesneimog/py4pd/blob/f304a093bf1012b6e16cbed8b8d3a3103f32b2b1/src/utils.cpp#L1274-L1278

Portanto, nesse código que me mandou ter ou não o `testeNone2.ignore_none=True` não vai fazer diferença. Você pode comentar ele.


``` py
def py4pdLoadObjects():
    testeNone = pd.new_object("teste.none")
    testeNone.addmethod_anything(teste_None)
    testeNone.py_out=True
    testeNone.add_object()

    testeNone2 = pd.new_object("teste.none2")
    #testeNone2.ignore_none=True 
    testeNone2.addmethod_anything(teste_None)
    testeNone2.add_object()

    testeNone3 = pd.new_object("teste.none3")
    testeNone3.addmethod_anything(teste_None)
    testeNone3.add_object()
```

Mas agora você falando não sei se isso é melhor mesmo, essa questão de converção de dados é complicada e meio arbitraria. Baseado nas suas perguntas, acho que é melhor colocar um 0 em None mesmo, né? E deixar o `ignore_none` sempre controlar de mandar ou não o 0, mesmo quando `py_true=False`. Sacou? O que acha?

Segunda coisa, agora não entendo de onde vinha esse 0 que vc tava tendo. `py_true` era `False` or `True`?

**Reply by emviveros** [reply]
Eu não utilizava py_true. Também eu havia tinha um caso onde eu tinha um retorno que me atrapalhava, mas quando vim revisitar o patch e código, estava resolvido. É provável que quando eu organizei para seguir na outra semana eu tenha feito algum ajuste que resolveu o problema. Talvez quando eu tratei o retorno da API que estou consultando, resolvi o problema. Acho que isso pode ser sempre feito e não há a necessidade de alterar o Py4Pd por isso, a não ser que se esteja muito preocupado com a legibilidade do código escrito para ser usado com o Py4Pd.

Neste momento, acho que apenas documentar que `ignore_none=True` só funciona para pyObjects seria uma boa, já que só funciona quando `py_out=True`.

Faz sentido pra você? Se precisar que eu destrinche qualquer coisa neste caso pode falar que posso discorrer mais se necessário.

**Reply by charlesneimog** [reply]
Acho que é isso, vou documentar isso e também documentar que quando `py_out=False` não haverá retorno em `None`.
